vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
+    template <typename T>
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
```

Comment:
Index is the current index into the buffer.

Reply:
and the index is implicitly incremented I see.
would be a good comment in the code :)

ok if added to the code as comment

Reply:
Will add a comment for this.

Reply:
Instead of writing a comment here, I opted to rename index_ to bufferIndex_.

The real problem here was that index_ is too generic a name, so it was unclear from the context what it was an index *for*. Giving it a more meaningful name like bufferIndex_ should go a long way in making the code self-documenting (which, in my opinion, should be preferred over comments when possible).